File: backend/src/main/java/com/taskflow/kanban/board/service/BoardService.java
package com.taskflow.kanban.board.service;

import com.taskflow.kanban.board.dto.BoardCreateDto;
import com.taskflow.kanban.board.dto.BoardDto;
import com.taskflow.kanban.board.dto.BoardUpdateDto;
import com.taskflow.kanban.board.entity.BoardRole;

import java.util.List;
import java.util.UUID;

public interface BoardService {
    BoardDto createBoard(BoardCreateDto createDto);
    BoardDto getBoardById(UUID id);
    void requireBoardAccess(UUID boardId);
    List<BoardDto> getBoardsByWorkspace(UUID workspaceId);
    List<BoardDto> getAllBoardsForCurrentUser();
    BoardDto updateBoard(UUID id, BoardUpdateDto updateDto);
    void deleteBoard(UUID id);
    void addMember(UUID boardId, UUID userId, BoardRole role);
    void removeMember(UUID boardId, UUID userId);
    void updateMemberRole(UUID boardId, UUID userId, BoardRole role);
}

File: frontend/src/app/core/services/board.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Board, BoardCreateDto, BoardUpdateDto } from '../models/board.model';
import { BoardRole, BoardMemberUpdateDto } from '../models/board-member.model';
import { Card } from '../models/card.model';
import { environment } from '../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class BoardService {
  private readonly API_URL = `${environment.apiUrl}/boards`;

  constructor(private http: HttpClient) {}

  getBoardsByWorkspace(workspaceId: string): Observable<Board[]> {
    const params = new HttpParams().set('workspaceId', workspaceId);
    return this.http.get<Board[]>(this.API_URL, { params });
  }

  /** All boards from every workspace the user is a member of (tasks shared across project) */
  getAllBoards(): Observable<Board[]> {
    return this.http.get<Board[]>(this.API_URL);
  }

  getBoardById(id: string): Observable<Board> {
    return this.http.get<Board>(`${this.API_URL}/${id}`);
  }

  getArchivedCardsByBoard(boardId: string): Observable<Card[]> {
    return this.http.get<Card[]>(`${this.API_URL}/${boardId}/archived-cards`);
  }

  createBoard(board: BoardCreateDto): Observable<Board> {
    return this.http.post<Board>(this.API_URL, board);
  }

  updateBoard(id: string, board: BoardUpdateDto): Observable<Board> {
    return this.http.put<Board>(`${this.API_URL}/${id}`, board);
  }

  deleteBoard(id: string): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  addMember(boardId: string, userId: string): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/${boardId}/members/${userId}`, {});
  }

  removeMember(boardId: string, userId: string): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${boardId}/members/${userId}`);
  }

  updateMemberRole(boardId: string, userId: string, role: BoardRole): Observable<void> {
    const updateDto: BoardMemberUpdateDto = { role };
    return this.http.put<void>(`${this.API_URL}/${boardId}/members/${userId}`, updateDto);
  }
}